src/webgpu/shader/execution/expression/constructor/non_zero.spec.ts

```diff
@@ -0,0 +1,419 @@
+          expected: matrixType.create(elements),
+        },
+      ]
+    );
+  });
+
+g.test('matrix_column_vectors')
+  .specURL('https://www.w3.org/TR/WGSL/#value-constructor-builtin-function')
+  .desc(`Test that a matrix constructed from column vectors produces the expected value`)
+  .params(u =>
+    u
+      .combine('inputSource', allInputSources)
+      .combine('type', ['f32', 'f16'] as const)
+      .combine('columns', [2, 3, 4] as const)
+      .combine('rows', [2, 3, 4] as const)
+      .combine('infer_type', [false, true] as const)
+  )
+  .beforeAllSubcases(t => {
+    if (t.params.type === 'f16') {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(async t => {
+    const elementType = Type[t.params.type];
+    const columnType = Type.vec(t.params.rows, elementType);
+    const matrixType = Type.mat(t.params.columns, t.params.rows, elementType);
+    const elements: number[] = [];
+    const columnVectors: VectorValue[] = [];
+    for (let column = 0; column < t.params.columns; column++) {
+      const columnElements: number[] = [];
+      for (let row = 0; row < t.params.rows; row++) {
+        const v = (column + 1) * 10 + (row + 1);
+        elements.push(v);
+        columnElements.push(v);
+      }
+      columnVectors.push(columnType.create(columnElements));
+    }
+    const fn = t.params.infer_type ? `mat${t.params.columns}x${t.params.rows}` : `${matrixType}`;
+    await run(
+      t,
+      basicExpressionBuilder(ops => `${fn}(${ops.join(', ')})`),
+      columnVectors.map(v => v.type),
+      matrixType,
+      t.params,
+      [
+        {
+          input: columnVectors,
+          expected: matrixType.create(elements),
+        },
+      ]
+    );
+  });
+
+g.test('array_elements')
+  .specURL('https://www.w3.org/TR/WGSL/#value-constructor-builtin-function')
+  .desc(`Test that a array constructed from element values produces the expected value`)
+  .params(u =>
+    u
+      .combine('inputSource', allInputSources)
+      .combine('type', ['bool', 'i32', 'u32', 'f32', 'f16', 'vec3f', 'vec4i'] as const)
+      .combine('length', [1, 5, 10] as const)
+      .combine('infer_type', [false, true] as const)
+  )
+  .beforeAllSubcases(t => {
+    if (t.params.type === 'f16') {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(async t => {
+    const elementType = Type[t.params.type];
+    const arrayType = Type.array(t.params.length, elementType);
+    const elements: number[] = [];
+    for (let i = 0; i < t.params.length; i++) {
+      elements.push((i + 1) * 10);
+    }
+    const fn = t.params.infer_type ? `array` : `${arrayType}`;
+    await run(
+      t,
+      basicExpressionBuilder(ops => `${fn}(${ops.join(', ')})`),
+      elements.map(e => elementType),
+      arrayType,
+      t.params,
+      [
+        {
+          input: elements.map(e => elementType.create(e)),
+          expected: arrayType.create(elements),
+        },
+      ]
+    );
+  });
```

Comment:
Please add .unimplemented() tests for structures.

Reply:
Done